File: partition.py
def partition(num, mem):
    result = 0
    for i in range(1, num + 1):
        gokaku = [i * (3 * i - 1) // 2, i * (3 * i + 1) // 2]
        if num < gokaku[0] and num < gokaku[1]:
            break
        for g in gokaku:
            if num >= g:
                if i % 2:
                    result += mem[num - g]
                else:
                    result -= mem[num - g]
    return result


def calcPartition(num):
    result = [1]
    for i in range(1, num + 1):
        result.append(partition(i, result))
    return result

num = int(input())
print(calcPartition(num)[-1])

File: partition.c
#include <stdio.h>
#define DATA_MAX 10000001
#define MOD 998244353

int partition(int num, int data[]) {
    int result = 0, i;
    int g1, g2;
    for (i = 1; i <= num; i++) {
        g1 = (i * (3 * i - 1)) / 2;
        g2 = (i * (3 * i + 1)) / 2;
        if (num < g1 && num < g2)
            break;
        if (num >= g1)
            result += (i%2)?data[num - g1]: - data[num - g1];
        if (num >= g2) result += (i%2)?data[num - g2]: - data[num - g2];
    }
    return result;
}

int calcPartition(int num) {
    int data[DATA_MAX] = {0};
    int i;
    data[0] = 1;
    for (i = 1; i <= num; i++)
        data[i] = partition(i, data);
    return data[num];
}

int main() {
    int N, tmp;
    tmp = scanf("%d", &N);
    printf("%d\n", calcPartition(N));
    return 0;
}